rust/yarp/build.rs

```diff
@@ -0,0 +1,654 @@
+    String,
+
+    #[serde(rename = "constant")]
+    Constant,
+
+    #[serde(rename = "constant[]")]
+    ConstantList,
+
+    #[serde(rename = "location")]
+    Location,
+
+    #[serde(rename = "location?")]
+    OptionalLocation,
+
+    #[serde(rename = "location[]")]
+    LocationList,
+
+    #[serde(rename = "uint32")]
+    UInt32,
+
+    #[serde(rename = "flags")]
+    Flags
+}
+
+#[derive(Debug, Deserialize)]
+struct NodeParam {
+    name: String,
+    r#type: NodeParamType,
+    kind: Option<String>
+}
+
+#[derive(Debug, Deserialize)]
+struct Node {
+    name: String,
+
+    #[serde(default)]
+    child_nodes: Vec<NodeParam>,
+
+    comment: String
+}
+
+#[derive(Debug, Deserialize)]
+struct Config {
+    nodes: Vec<Node>
+}
+
+/// The main function for the build script. This will be run by Cargo when
+/// building the library.
+/// 
+fn main() -> Result<(), Box<dyn std::error::Error>> {
+    let file = std::fs::File::open(config_path())?;
+    let config: Config = serde_yaml::from_reader(file)?;
+
+    write_bindings(&config)?;
+    Ok(())
+}
+
+/// Gets the path to the config.yml file at `[root]/config.yml`.
+///
+fn config_path() -> PathBuf {
+    cargo_manifest_path()
+        .join("../../config.yml")
+        .canonicalize()
+        .unwrap()
+}
+
+/// Gets the path to the root directory at `[root]`.
+/// 
+fn cargo_manifest_path() -> PathBuf {
+    PathBuf::from(std::env::var_os("CARGO_MANIFEST_DIR").unwrap())
+}
+
+/// Returns the name of a C struct from the given node name.
+fn struct_name(name: &str) -> String {
+    let mut result = String::with_capacity(1 + name.len());
+
+    for char in name.chars() {
+        if char.is_uppercase() {
+            result.push('_');
+        }
+        result.push(char.to_lowercase().next().unwrap());
+    }
+
+    result
+}
+
+/// Returns the name of the C type from the given node name.
+fn type_name(name: &str) -> String {
+    let mut result = String::with_capacity(8 + name.len());
+    result.push_str("YP_NODE");
+
+    for char in name.chars() {
+        if char.is_uppercase() {
+            result.push('_');
+        }
+        result.push(char.to_uppercase().next().unwrap());
+    }
+
+    result
+}
+
+/// Write the generated struct for the node to the file.
+fn write_node(file: &mut File, node: &Node) -> Result<(), Box<dyn std::error::Error>> {
+    let mut example = false;
+
+    for line in node.comment.lines() {
+        if let Some(stripped) = line.strip_prefix("    ") {
+            if !example {
+                writeln!(file, "/// ```ruby")?;
+                example = true;
+            }
+            writeln!(file, "/// {}", stripped)?;
+        } else {
+            if example {
+                writeln!(file, "/// ```")?;
+                example = false;
+            }
+            writeln!(file, "/// {}", line)?;
+        }
+    }
+
+    if example {
+        writeln!(file, "/// ```")?;
+    }
+
+    writeln!(file, "pub struct {}<'pr> {{", node.name)?;
```

Comment:
Any reason to not use ERB templating for this like for all other languages?
I was thinking I could peek at the generated code for Rust using `bundle exec rake templates` but no luck.
Also this would mean we likely need to modify here and `bin/template` if e.g. the format of `config.yml` changes.
That might be a barrier, e.g. my Rust is quite rusty (OTOH I have no intentions to change the format of config.yml).

Reply:
Because when we publish to `crates.io`, this means we only need a copy of `config.yml` and don't have to have all of the mechanisms to template. If `config.yml` changes I'm happy to update it here, it's not hard to change. Plus it will fail compilation, which is a nice perk.